zkevm-circuits/src/keccak_circuit/keccak_padding_multirows_2.rs

```diff
@@ -0,0 +1,637 @@
+    pub(crate) acc_len: u32,
+    pub(crate) block_rows: [KeccakPaddingBlockRow<F>; KECCAK_REGION_HEIGHT as usize],
+}
+
+#[derive(Debug)]
+pub(crate) struct KeccakPaddingBlockRow<F: Field> {
+    pub(crate) curr_padding_sum: u32,
+    pub(crate) d_bits: [u8; KECCAK_REGION_WIDTH],
+    pub(crate) d_lens: [u32; KECCAK_REGION_WIDTH_IN_BYTES],
+    pub(crate) d_rlcs: [F; KECCAK_REGION_WIDTH_IN_BYTES],
+    pub(crate) s_flags: [bool; KECCAK_REGION_WIDTH_IN_BYTES],
+}
+
+/// KeccakPaddingCircuit
+#[derive(Default)]
+pub struct KeccakPaddingCircuit<F: Field> {
+    inputs: Vec<KeccakPaddingBlock<F>>,
+    size: usize,
+    _marker: PhantomData<F>,
+}
+
+impl<F: Field> KeccakPaddingCircuit<F> {
+    fn r() -> F {
+        F::from(123456)
+    }
+}
+
+impl<F: Field> Circuit<F> for KeccakPaddingCircuit<F> {
+    type Config = KeccakPaddingConfig<F>;
+    type FloorPlanner = SimpleFloorPlanner;
+
+    fn without_witnesses(&self) -> Self {
+        Self::default()
+    }
+
+    fn configure(meta: &mut ConstraintSystem<F>) -> Self::Config {
+        KeccakPaddingConfig::configure(meta)
+    }
+
+    fn synthesize(&self, config: Self::Config, layouter: impl Layouter<F>) -> Result<(), Error> {
+        config.assign(
+            layouter,
+            self.size,
+            &self.inputs[0],
+            KeccakPaddingCircuit::r(),
+        )?;
+        Ok(())
+    }
+}
+
+impl<F: Field> KeccakPaddingConfig<F> {
+    pub(crate) fn configure(meta: &mut ConstraintSystem<F>) -> Self {
+        let q_enable = meta.selector();
+        let q_end = meta.advice_column();
+        let is_first_row = meta.fixed_column();
+        let is_last_row = meta.fixed_column();
+        let curr_padding_sum = meta.advice_column();
+        let d_bits = [(); KECCAK_REGION_WIDTH].map(|_| meta.advice_column());
+        let d_lens = [(); KECCAK_REGION_WIDTH_IN_BYTES].map(|_| meta.advice_column());
+        let d_rlcs = [(); KECCAK_REGION_WIDTH_IN_BYTES].map(|_| meta.advice_column());
+        let s_flags = [(); KECCAK_REGION_WIDTH_IN_BYTES].map(|_| meta.advice_column());
+        let randomness = meta.advice_column();
+
+        meta.create_gate("prev should be 0 for the 1st row", |meta| {
+            let mut cb = BaseConstraintBuilder::new(5);
+
+            // len & rlc are passed down by previous circuit, they are not necessarily 0.
+            let is_first = meta.query_fixed(is_first_row, Rotation::cur());
+
+            cb.condition(is_first.clone(), |cb| {
+                cb.require_zero(
+                    "prev_s_flag == 0",
+                    meta.query_advice(s_flags[KECCAK_REGION_WIDTH_IN_BYTES - 1], Rotation::prev()),
+                );
+            });
+            cb.condition(is_first.clone(), |cb| {
+                cb.require_zero(
+                    "prev_padding_sum == 0",
+                    meta.query_advice(curr_padding_sum, Rotation::prev()),
+                );
+            });
+
+            cb.gate(meta.query_selector(q_enable))
```

Comment:
All constraints already have the same `is_first` selector here so instead of `q_enable` you can just use `is_first` and don't do the `condition`s. `is_first` is also a fixed column (like `q_enable`) so you can use that as the main selector directly (this doesn't work for selectors in advice columns).